2 - cluster organisations and link them to the STM topics/preprocessing.py

```diff
@@ -0,0 +1,273 @@
+    global ngram_replacements
+    
+    datafr = pd.read_csv(os.path.join(data_dir, "master-ner-results.csv"))
+    # copy data
+    new_data = {}
+    keys =[]
+    for key in data:
+        # keys.append(key)
+        # print(key, ' - ', type(key))
+        new_data[str(lookup_correct_docid(key))] = data[key]
+        
+    # keys.sort()
+    # print(keys)
+    # replace phrases with unigrams
+    entity_dict = {}
+    
+    # print(len(new_data))
+    
+    for k in new_data:
+        new_data[k] = preprocess_ner(new_data[k], stopwords=None)
+        
+        # first filter the rows pertaining to the given docid
+        docid_datafr = datafr[datafr['docid'] == int(k)]
+        # print('frame size: ', len(docid_datafr))
+        docid_org_per_datafr = docid_datafr[docid_datafr['entity_type'].isin(['ORG', 'PERSON', 'NORP', 'LOC', 'FAC', 'GPE'])]
+        # loop through each entity mention (row) in the dataframe
+        docid_org_per_datafr = docid_org_per_datafr.reset_index()  # make sure indexes pair with number of rows
+        unique_sentences = pd.unique(docid_org_per_datafr['sentence'])
+
+        for sentence in unique_sentences:
+            curr_sent_df = docid_org_per_datafr[docid_org_per_datafr['sentence'] == sentence]
+            curr_sent_entities = curr_sent_df['entity'].tolist()
+            replace_patterns = {}
+            for entity in curr_sent_entities:
+                if isinstance(entity, str):
+                    if entity[0:4] == 'the ':
+                        entity = entity[4:]
+                    named_entity_tokens = entity.strip().replace('"', '').replace("'", '').replace(",",'').replace('’','').replace('‘','').replace('“','').replace('”','').split()
+                    if (len(named_entity_tokens) > 1):
+                        # form single token from multiple ones
+                        single_token_entity = '_'.join(named_entity_tokens)
+                        replace_patterns[entity] = single_token_entity
+                        entity_dict[entity] = single_token_entity
+                        # print('entity: ', entity, ' - ', 'replacement: ', entity_dict[entity])
+
+            new_sentence = replace_all(sentence, replace_patterns)
+            new_data[k] = new_data[k].replace(sentence, new_sentence)
+            
+    # print('entity_dict size: ', len(entity_dict))
+
+    with open(os.path.join(data_dir, "entity_dict.json"), 'w') as fp:
+        json.dump(entity_dict, fp)
+
+    # Declare a list that is to be converted into a column
+    single_tokens = []
+
+    for item in datafr['entity'].tolist():
+        if isinstance(item, str):
+            if item[:4] == 'the ':
+                item = item[4:]
+
+            named_entity_tokens = item.strip().replace('"','').replace("'",'').replace(",",'').replace('’','').replace('‘','').replace('“','').replace('”','').split()
+            
+            if (len(named_entity_tokens) > 1):
+                single_tokens.append(entity_dict[item])
```

Comment:
```suggestion
            if (len(named_entity_tokens) > 1 and item in entity_dict):
                single_tokens.append(entity_dict[item])
```

I would add this as a safeguard in case the NER results are not entirely deterministic. For me, it threw a KeyError here because "dominican republic" was not a key in `enitity_dict`.

Reply:
okay, updated.

> I managed to run all the notebooks now, so here are a few general comments:
> 
> * The code contains still quite a few commented out sections. Perhaps we should remove some of them before we wrap up?
> * Some cells of the notebook take quite long to run. A note on that would probably be useful!
> * The code still contains a few duplicated functions which could be put into one single location (I pointed out where).

I addressed the first two points I think. The duplicated functions thing I did not get to yet because it requires a little more time and perhaps tricky changes but I will try to get to it soon,